File: src/ee/ads/internal/AdViewHelper.cpp
//
//  AdViewHelper.cpp
//  ee_x
//
//  Created by Zinge on 10/12/17.
//
//

#include "ee/ads/internal/AdViewHelper.hpp"

#include <ee/nlohmann/json.hpp>

#include <ee/core/Utils.hpp>
#include <ee/core/internal/IMessageBridge.hpp>

namespace ee {
namespace ads {
using Self = AdViewHelper;

Self::AdViewHelper(IMessageBridge& bridge, const MessageHelper& helper)
    : bridge_(bridge)
    , helper_(helper)
    , anchorX_(0.0f)
    , anchorY_(0.0f) {}

bool Self::isLoaded() const {
    assert(isMainThread());
    auto response = bridge_.call(helper_.isLoaded());
    return core::toBool(response);
}

void Self::load() {
    assert(isMainThread());
    bridge_.call(helper_.load());
}

std::pair<float, float> Self::getAnchor() const {
    return std::pair(anchorX_, anchorY_);
}

void Self::setAnchor(float x, float y) {
    assert(isMainThread());
    auto [width, height] = getSize();
    auto [positionX, positionY] = getPositionTopLeft();
    setPositionTopLeft(positionX - static_cast<int>((x - anchorX_) * width),
                       positionY - static_cast<int>((y - anchorY_) * height));
    anchorX_ = x;
    anchorY_ = y;
}

std::pair<int, int> Self::getPosition() const {
    assert(isMainThread());
    auto [width, height] = getSize();
    auto [anchorX, anchorY] = getAnchor();
    auto [x, y] = getPositionTopLeft();
    return std::pair(x + anchorX * width, y + anchorY * height);
}

void Self::setPosition(int x, int y) {
    assert(isMainThread());
    auto [width, height] = getSize();
    auto [anchorX, anchorY] = getAnchor();
    setPositionTopLeft(x - static_cast<int>(anchorX * width),
                       y - static_cast<int>(anchorY * height));
}

std::pair<int, int> Self::getPositionTopLeft() const {
    assert(isMainThread());
    auto response = bridge_.call(helper_.getPosition());
    auto json = nlohmann::json::parse(response);
    auto x = json["x"].get<int>();
    auto y = json["y"].get<int>();
    return std::pair(x, y);
}

void Self::setPositionTopLeft(int x, int y) {
    assert(isMainThread());
    nlohmann::json json;
    json["x"] = x;
    json["y"] = y;
    bridge_.call(helper_.setPosition(), json.dump());
}

std::pair<int, int> Self::getSize() const {
    assert(isMainThread());
    auto response = bridge_.call(helper_.getSize());
    auto json = nlohmann::json::parse(response);
    auto width = json["width"].get<int>();
    auto height = json["height"].get<int>();
    return std::make_pair(width, height);
}

void Self::setSize(int width, int height) {
    assert(isMainThread());
    auto [currentWidth, currentHeight] = getSize();
    auto [anchorX, anchorY] = getAnchor();
    auto [x, y] = getPositionTopLeft();
    setPositionTopLeft(
        x - static_cast<int>((width - currentWidth) * anchorX),
        y - static_cast<int>((height - currentHeight) * anchorY));

    nlohmann::json json;
    json["width"] = width;
    json["height"] = height;
    bridge_.call(helper_.setSize(), json.dump());
}

bool Self::isVisible() const {
    assert(isMainThread());
    auto response = bridge_.call(helper_.isVisible());
    return core::toBool(response);
}

void Self::setVisible(bool visible) {
    assert(isMainThread());
    bridge_.call(helper_.setVisible(), core::toString(visible));
}
} // namespace ads
} // namespace ee

File: src/ee/ads/internal/AdViewHelper.swift
//
//  AdViewHelper.swift
//  ee-x-366f64ba
//
//  Created by eps on 6/24/20.
//

import Foundation

class AdViewHelper {
    private let _bridge: IMessageBridge
    private var _view: IAdView
    private let _helper: MessageHelper

    init(_ bridge: IMessageBridge,
         _ view: IAdView,
         _ helper: MessageHelper) {
        _bridge = bridge
        _view = view
        _helper = helper
    }

    func registerHandlers() {
        _bridge.registerHandler(_helper.isLoaded) { _ in
            Utils.toString(self._view.isLoaded)
        }
        _bridge.registerHandler(_helper.load) { _ in
            self._view.load()
            return ""
        }
        _bridge.registerHandler(_helper.getPosition) { _ in
            let position = self._view.position
            return EEJsonUtils.convertDictionary(toString: [
                "x": position.x,
                "y": position.y
            ])
        }
        _bridge.registerHandler(_helper.setPosition) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let x = dict["x"] as? Int else {
                assert(false, "Invalid argument")
                return ""
            }
            guard let y = dict["y"] as? Int else {
                assert(false, "Invalid argument")
                return ""
            }
            self._view.position = CGPoint(x: x, y: y)
            return ""
        }
        _bridge.registerHandler(_helper.getSize) { _ in
            let size = self._view.size
            return EEJsonUtils.convertDictionary(toString: [
                "width": size.width,
                "height": size.height
            ])
        }
        _bridge.registerHandler(_helper.setSize) { message in
            let dict = EEJsonUtils.convertString(toDictionary: message)
            guard let width = dict["width"] as? Int else {
                assert(false, "Invalid argument")
                return ""
            }
            guard let height = dict["height"] as? Int else {
                assert(false, "Invalid argument")
                return ""
            }
            self._view.size = CGSize(width: width, height: height)
            return ""
        }
        _bridge.registerHandler(_helper.isVisible) { _ in
            Utils.toString(self._view.isVisible)
        }
        _bridge.registerHandler(_helper.setVisible) { message in
            self._view.isVisible = Utils.toBool(message)
            return ""
        }
    }

    func deregisterHandlers() {
        _bridge.deregisterHandler(_helper.isLoaded)
        _bridge.deregisterHandler(_helper.load)
        _bridge.deregisterHandler(_helper.getPosition)
        _bridge.deregisterHandler(_helper.setPosition)
        _bridge.deregisterHandler(_helper.getSize)
        _bridge.deregisterHandler(_helper.setSize)
        _bridge.deregisterHandler(_helper.isVisible)
        _bridge.deregisterHandler(_helper.setVisible)
    }
}
